_posts/2024-04-09-flake-schemas.md

```diff
@@ -0,0 +1,143 @@
+
+With that out of the way, we can now define the schema inputs in our flake:
+
+{% highlight nix %}
+{
+  inputs = {
+    nixpkgs.url = "nixpkgs/nixos-unstable";
+    flake-schemas.url = "github:gvolpe/flake-schemas";
+
+    ## nix client with schema support: see https://github.com/NixOS/nix/pull/8892
+    nix-schema = {
+      inputs.flake-schemas.follows = "flake-schemas";
+      url = "github:DeterminateSystems/nix-src/flake-schemas";
+    };
+  };
+}
+{% endhighlight %}
+
+And as outputs, we define the standard schemas plus a custom one we'll see soon:
+
+{% highlight nix %}
+{
+  schemas =
+    inputs.flake-schemas.schemas //
+    import ./lib/schemas.nix { inherit (inputs) flake-schemas; };
+}
+{% endhighlight %}
+
+As pointed out by Greg in [one of the open issues](https://github.com/DeterminateSystems/flake-schemas/issues/14), the `nix` client from the fork doesn't build straight away, so we need some tweaks that the following overlay takes care of:
+
+{% highlight bash %}
+f: p: {
+  nix-schema = inputs.nix-schema.packages.${system}.nix.overrideAttrs (old: {
+    doCheck = false;
+    doInstallCheck = false;
+    postInstall = old.postInstall + ''
+      rm $out/bin/nix-*
+      mv $out/bin/nix $out/bin/nix-schema
+    '';
+  });
+};
+{% endhighlight %}
+
+Then we can add it to our system packages, conveniently named `nix-schema` to avoid conflicts:
+
+{% highlight nix %}
+{
+  environment.systemPackages = [
+    pkgs.nix-schema
+  ];
+}
+{% endhighlight %}
+
+NOTE: the `rm $out/bin/nix-*` bit ensures that all the usual commands such as `nix-build`, `nix-shell`, etc don't get overridden, which is [one of the issues I got myself into](https://discourse.nixos.org/t/the-program-nix-build-is-not-in-your-path/42943).
+
+Next we have the `lib/schemas.nix` file referenced above, which defines the schema for the `out` custom output in my `nix-config` repo, which is a rather simple example of a flake schema:
+
+{% highlight nix %}
+{ flake-schemas }:
+
+{
+  out = {
+    version = 1;
+    doc = "Exports custom attrsets...";
+    inventory = output:
+      flake-schemas.lib.mkChildren (builtins.mapAttrs
+        (_: _: {
+          what = "custom instance to be used by consumers of this flake";
+        })
+        output);
+  };
+}
+{% endhighlight %}
+
+I've also handily exposed the `nix` client as a flake app:
+
+{% highlight nix %}
+{
+  apps.${system}."nix" = {
+    type = "app";
+    program = "${pkgs.nix-schema}/bin/nix-schema";
+  };
+}
+{% endhighlight %}
+
+So we can run it directly from this flake without installing the special `nix` version.
+
+{% highlight bash %}
+$ nix run github:gvolpe/nix-config#nix -- --version
+nix-schema (Nix) 2.21.0pre20240311_d76e5fb
+{% endhighlight %}
+
+### Give it a try
+
+You can run the following command in any of your flake repositories to see the new "show" output:
+
+{% highlight bash %}
+$ nix run github:gvolpe/nix-config#nix -- flake show
```

Comment:
**nitpick:** sadly this doesn't work on aarch64-darwin:

```
error: flake 'github:gvolpe/nix-config' does not provide attribute 'apps.aarch64-darwin.nix', 'packages.aarch64-darwin.nix', 'legacyPackages.aarch64-darwin.nix' or 'nix'
```

Reply:
ah that's because my flake only exposes the `nix` client compiled for Linux, should work on Darwin too if you compile the client too (I won't do it, takes too long 😄 )

Reply:
I wouldn't mind rebuilding it, but without you exposing it I don't think I can easily schedule a build 😬